src/main/java/com/faforever/client/patch/SimpleHttpFeaturedModUpdaterTask.java

```diff
@@ -54,32 +61,46 @@ protected PatchResult call() throws Exception {
     updateMessage(i18n.get("updater.readingFileList"));
 
     List<FeaturedModFile> featuredModFiles = fafService.getFeaturedModFiles(featuredMod, version).get();
+    Path fafDataDirectory = preferencesService.getFafDataDirectory();
 
     Path initFile = null;
+    Path cacheFilePath;
+    Path targetPath;
+
+    Map<String, String> knownTargetHashes = new HashMap<>();
+
+    // Download to cache if file exists in target place, otherwise, download to direct place.
     for (FeaturedModFile featuredModFile : featuredModFiles) {
-      Path fafDataDirectory = preferencesService.getFafDataDirectory();
-      Path targetPath = fafDataDirectory
-          .resolve(featuredModFile.getGroup())
-          .resolve(featuredModFile.getName());
-
-      if (Files.exists(targetPath)
-          && featuredModFile.getMd5().equals(com.google.common.io.Files.hash(targetPath.toFile(), Hashing.md5()).toString())) {
-        logger.debug("Already up to date: {}", targetPath);
-      } else {
-        Files.createDirectories(targetPath.getParent());
-        updateMessage(i18n.get("updater.downloadingFile", targetPath.getFileName()));
-
-        String url = featuredModFile.getUrl();
-        downloadService.downloadFile(new URL(url), targetPath, this::updateProgress);
-        UpdaterUtil.extractMoviesIfPresent(targetPath, fafDataDirectory);
+      targetPath = fafDataDirectory.resolve(featuredModFile.getGroup()).resolve(featuredModFile.getName());
+      cacheFilePath = featuredModFileCacheService.getCachedFilePath(featuredModFile);
+
+      String existingTargetFileHash = null;
+      if (Files.exists(targetPath)) {
+        existingTargetFileHash = com.google.common.io.Files.hash(targetPath.toFile(), Hashing.md5()).toString();
```

Comment:
maybe static import

Reply:
`import java.nio.file.Files;` vs `com.google.common.io.Files`, it will fight each other, how to solve such?

Reply:
static import just the function